state/traverse.go

```diff
@@ -7,27 +7,59 @@ import (
 	"github.com/onflow/flow-go/storage"
 )
 
-// Traverse traverses a chain segment beginning with the start block (inclusive)
-// and ending with the end block (inclusive). Blocks are traversed in reverse
+// functor that will be get called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) (bool, error)
+
+// TraverseBackward traverses a chain segment beginning with the start block (inclusive)
+// Blocks are traversed in reverse
 // height order, meaning the end block must be an ancestor of the start block.
 // The callback is called for each block in this segment.
-func Traverse(headers storage.Headers, start, end flow.Identifier, callback func(header *flow.Header) error) error {
-
-	nextID := start
+// Return value of callback is used to decide if it should continue or not.
+func TraverseBackward(headers storage.Headers, startBlockID flow.Identifier, visitor onVisitBlock) error {
+	blockID := startBlockID
 	for {
-		// retrieve the next block in the segment and pass it to the callback
-		next, err := headers.ByBlockID(nextID)
+		block, err := headers.ByBlockID(blockID)
 		if err != nil {
-			return fmt.Errorf("could not get segment block (id=%x): %w", nextID, err)
+			return fmt.Errorf("could not get block header (%x): %w", blockID, err)
+		}
+
+		shouldContinue, err := visitor(block)
+		if !shouldContinue {
+			break
 		}
-		err = callback(next)
 		if err != nil {
-			return fmt.Errorf("error in callback: %w", err)
+			return err
+		}
+		blockID = block.ParentID
+	}
+	return nil
+}
+
+// TraverseForward traverses a chain segment in forward order.
+// Implements a recursive descend starting at the `forkHead` towards the genesis block. The descend continues as long as `shouldContinue` returns true. All visited blocks, which `shouldContinue` returned true for, are fed into the `visitor` callback starting from the block with the lowest height, in order of increasing height. The last block that is fed into `visitor` is the `forkHead`.
```

Comment:
not recursive anymore

Reply:
maybe we could improve the input name or the comments orherwise sounds like travsersing from the forkHead forward

Reply:
```suggestion
// It first traverses backward with the given `forkHead` and `shouldContinue`. Once stopped, it will pass the visited blocks to the given `visitor` in a forward order. The last block that is fed into `visitor` is the `forkHead`.
```